src/main/java/uk/ac/soton/itinnovation/security/modelvalidator/attackpath/RecommendationsAlgorithm.java

```diff
@@ -0,0 +1,490 @@
+
+        checkRequestedRiskCalculationMode(riskCalculationMode);
+
+        return calculateAttackTreeInternal(targetUris, riskCalculationMode, allPaths);
+    }
+
+    private AttackTree calculateAttackTreeInternal(List<String> targetUris, String riskCalculationMode, boolean singlePath)
+            throws RuntimeException {
+
+        boolean isFutureRisk = apd.isFutureRisk(riskCalculationMode);
+        AttackTree attackTree = null;
+
+        try {
+            final long startTime = System.currentTimeMillis();
+            attackTree = new AttackTree(targetUris, isFutureRisk, singlePath, apd);
+            attackTree.stats();
+            final long endTime = System.currentTimeMillis();
+            logger.info("AttackPathAlgorithm.calculateAttackTree: execution time {} ms", endTime - startTime);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        return attackTree;
+    }
+
+    /**
+     * Build CSG recommendations tree
+     * @param le
+     * @param myNode
+     * @return 
+     */
+    private CSGNode applyCSGs(LogicalExpression le, CSGNode myNode) {
+        logger.debug("applyCSGs() recursive method");
+
+        if (myNode == null) {
+            myNode = new CSGNode();
+        }
+
+        // convert LE to DNF
+        le.applyDNF(300);
+
+        // convert from CSG logical expression to list of CSG options
+        List<Expression> csgOptions = le.getListFromOr();
+
+        logger.debug("Derived DNF (OR) CSG expressions: {} (options).", csgOptions.size());
+        for (Expression csgOption : csgOptions) {
+            logger.debug("   └──> {}", csgOption);
+        }
+
+        // examine CSG options
+        for (Expression csgOption : csgOptions) {
+            logger.debug("examining CSG LE option: {}", csgOption);
+
+            List<Variable> options = le.getListFromAnd(csgOption);
+
+            List<String> csgList = new ArrayList<>();
+
+            for (Variable va : options) {
+                csgList.add(va.toString());
+            }
+            logger.debug("CSG flattened list: {}", csgList);
+
+            logger.debug("adding a child CSGNode with {} csgs", csgList.size());
+            CSGNode childNode = new CSGNode(csgList);
+            myNode.addChild(childNode);
+
+            Set<String> csSet = new HashSet<>();
+            for (String csg : csgList) {
+                for (String cs : apd.getCsgInactiveControlSets(csg)) {
+                    csSet.add(cs);
+                }
+            }
+
+            // store CS set in the node to reconstruct the final CS list
+            // correctly in the Recommendation report for nested iterations.
+            childNode.setCsList(csSet);
+
+            logger.debug("CS set for LE CSG_option {}", csgOption);
+            logger.debug("  └──> {}", csSet);
+
+            // apply all CS in the CS_set
+            if (csSet.isEmpty()) {
+                logger.debug("EMPTY csSet is found, skipping iteration");
```

Comment:
I think this should never happen. In which case logging at error level would make more sense